src/mapper.rs

```diff
@@ -0,0 +1,153 @@
+                    }
+                    class = ClassMapping {
+                        original,
+                        obfuscated,
+                        members: HashMap::new(),
+                    }
+                }
+                MappingRecord::Member {
+                    original,
+                    obfuscated,
+                    original_class,
+                    line_mapping,
+                    ..
+                } => {
+                    if let Some(line_mapping) = line_mapping {
+                        if let Some(original_startline) = line_mapping.original_startline {
+                            let members = class.members.entry(obfuscated).or_insert_with(|| vec![]);
+                            members.push(MemberMapping {
+                                startline: line_mapping.startline,
+                                endline: line_mapping.endline,
+                                original_class,
+                                original,
+                                original_startline,
+                                original_endline: line_mapping.original_endline,
+                            });
+                        }
+                    }
+                }
+            }
+        }
+        if !class.original.is_empty() {
+            classes.insert(class.obfuscated, class);
+        }
+
+        Self { classes }
+    }
+
+    /// Remaps a single Stackframe.
+    ///
+    /// Returns one or more [`StackFrame`]s, based on the information in
+    /// the proguard mapping. This can return more than one frame in the case
+    /// of inlined functions. In that case, frames are sorted top to bottom.
+    ///
+    /// [`StackFrame`]: struct.StackFrame.html
+    pub fn remap_frame(
+        &'s self,
+        frame: &StackFrame<'s>,
+    ) -> impl Iterator<Item = StackFrame<'s>> + 's {
+        if let Some(class) = self.classes.get(frame.class.as_ref()) {
+            if let Some(members) = class.members.get(frame.method.as_ref()) {
+                // find matches based on line number
+                let mapped: Vec<_> = members
+                    .iter()
+                    .filter(|m| frame.line >= m.startline && frame.line <= m.endline)
+                    .map(|member| {
+                        // parents of inlined frames don’t have an `endline`, and
+                        // the top inlined frame need to be correctly offset.
+                        let line = if member.original_endline.is_none() {
+                            member.original_startline
+                        } else {
+                            member.original_startline + frame.line - member.startline
+                        };
+                        let file = member
+                            .original_class
+                            .map(|c| {
+                                let c = c.rsplit('.').next().unwrap();
+                                Cow::Owned(format!("{}.java", c))
```

Comment:
Note to self: Add a comment why this happens. Looks like proguard/retrace infers the filename from the class, in case this is a cross-class inline.